parser/src/aspects/if_else.rs

```diff
@@ -0,0 +1,224 @@
+use lexer::token::TokenType;
+use lexer::token::TokenType::{Else, NewLine, SemiColon};
+use crate::ast::Expr;
+use crate::moves::{MoveOperations, of_type, repeat, spaces};
+use crate::parser::{Parser, ParseResult};
+use crate::ast::flow_control::If;
+
+///parser aspect for if and else expressions.
+pub trait IfElseAspect<'a> {
+    ///parse a if with optional else expression.
+    /// `parse_branch` argument defines how the branches must be parsed.
+    fn parse_if<F>(&mut self, parse_branch: F) -> ParseResult<Expr<'a>>
+        where F: FnMut(&mut Self) -> ParseResult<Expr<'a>>;
+}
+
+impl<'a> IfElseAspect<'a> for Parser<'a> {
+    fn parse_if<F>(&mut self, mut parse_branch: F) -> ParseResult<Expr<'a>>
+        where F: FnMut(&mut Self) -> ParseResult<Expr<'a>> {
+        self.cursor.force(
+            of_type(TokenType::If),
+            "expected 'if' at start of if expression",
+        )?;
+        let condition = self.expression_statement()?;
+
+        //skip only one semicolon if any, surrounded by newlines and spaces
+        self.cursor.advance(spaces().then(
+            repeat(spaces().pipe(of_type(NewLine)))
+                .then(of_type(SemiColon))
+                .then(repeat(spaces().pipe(of_type(NewLine))))
+        ));
+
+        //the success_branch of the if
+        let success_branch = parse_branch(self)?;
+
+        //parse the 'else' branch.
+        let fail_branch =
+            if self.cursor.advance(
+                repeat(spaces().pipe(of_type(NewLine)))
+                    .then(of_type(SemiColon))
+                    .then(repeat(spaces().pipe(of_type(NewLine))))
+                    .then(of_type(Else))
```

Comment:
Couldn't this move be constructed from the previous move, to highlight that only the `.then(of_type(Else))` has been added?